File: risk-lib/src/main/java/net/mjduffin/risk/lib/entities/DiceManager.java
package net.mjduffin.risk.lib.entities;

import net.mjduffin.risk.lib.usecase.GameplayException;

import java.util.Collections;
import java.util.PriorityQueue;

public class DiceManager {
    private final DieThrow random;

    public static class Result {
        protected int attackers;
        protected int defenders;

        public int getAttackers() {
            return attackers;
        }

        public int getDefenders() {
            return defenders;
        }
    }

    public DiceManager(DieThrow random) {
        this.random = random;
    }

    public Result engage(int attackers, int defenders) throws GameplayException {
        if (attackers > 3 || defenders > 2) {
            throw new GameplayException("Incorrect number of dice");
        }
        PriorityQueue<Integer> attackResults = new PriorityQueue<>(Collections.reverseOrder());
        PriorityQueue<Integer> defendResults = new PriorityQueue<>(Collections.reverseOrder());

        for (int i = 0; i < attackers; i++) {
            int num = random.getDieValue();
            attackResults.add(num);
        }
        for (int i = 0; i < defenders; i++) {
            int num = random.getDieValue();
            defendResults.add(num);
        }

        Result result = new Result();
        result.attackers = attackers;
        result.defenders = defenders;
        while (!attackResults.isEmpty() && !defendResults.isEmpty()) {
            int a = attackResults.poll();
            int d = defendResults.poll();

            if (a > d) {
                result.defenders--;
            } else {
                result.attackers--;
            }
        }

        return result;
    }
}

File: risk-lib/src/main/java/net/mjduffin/risk/lib/entities/DiceManager.kt
package net.mjduffin.risk.lib.entities

import net.mjduffin.risk.lib.usecase.GameplayException
import java.util.*

class DiceManager(private val random: DieThrow) {
    data class Result(val attackers: Int = 0, val defenders: Int = 0)

    fun engage(attackers: Int, defenders: Int): Result {
        if (attackers > 3 || defenders > 2) {
            throw GameplayException("Incorrect number of dice")
        }
        val attackResults = PriorityQueue(Collections.reverseOrder<Int>())
        val defendResults = PriorityQueue(Collections.reverseOrder<Int>())
        for (i in 0 until attackers) {
            val num = random.dieValue
            attackResults.add(num)
        }
        for (i in 0 until defenders) {
            val num = random.dieValue
            defendResults.add(num)
        }
        var remainingAttackers = attackers
        var remainingDefenders = defenders
        while (!attackResults.isEmpty() && !defendResults.isEmpty()) {
            val a = attackResults.poll()
            val d = defendResults.poll()
            if (a > d) {
                remainingDefenders--
            } else {
                remainingAttackers--
            }
        }
        return Result(remainingAttackers, remainingDefenders)
    }
}